Crash at API::getContentRuleListSourceFromMappedFile().

xeenon opened the pull request:
#### fa5d8d81450ba9d0a3f29ddc100b7f2e80c1420b
<pre>
Crash at API::getContentRuleListSourceFromMappedFile().
<a href="https://webkit.org/b/283996">https://webkit.org/b/283996</a>
<a href="https://rdar.apple.com/140450689">rdar://140450689</a>

Reviewed by Alex Christensen.

Make the handling of getContentRuleListSourceFromMappedFile() more robust by checking header sizes
to the mapped data size. This was failing when the version got bumped and fallback to recompiling
from the original source was needed.

Improved mismatch version testing by dropping the version by one and keeping all the compiled data
the same to test proper version upgrades from the original source.

Added two corrupt size tests to error on invalid data sizes or empty data (missing source).

* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::getContentRuleListSourceFromMappedFile): Added logging and size checks.
(API::ContentRuleListStore::lookupContentRuleListFile): Check fileSize() against the mapped data.
(API::ContentRuleListStore::invalidateContentRuleListVersion): Changed to set the version back by one.
(API::ContentRuleListStore::corruptContentRuleListHeader): Added.
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStore.mm:
(-[WKContentRuleListStore _corruptContentRuleListHeaderForIdentifier:usingCurrentVersion:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStorePrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentExtensionStore.mm:
(TEST_F(WKContentRuleListStoreTest, VersionMismatch)): Updated expectations since source is preserved now.
(TEST_F(WKContentRuleListStoreTest, CorruptHeaderEmpty)): Added.
(TEST_F(WKContentRuleListStoreTest, CorruptHeaderRandom)): Added.

Canonical link: <a href="https://commits.webkit.org/287319@main">https://commits.webkit.org/287319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/483da79b00b9891bde34681253bf17c4e980a9b1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58173 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32516 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83772 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30341 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81274 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67273 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6439 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61923 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19839 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82208 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51970 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72117 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42227 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49321 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26237 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28712 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70437 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26662 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85162 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6452 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4466 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70170 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6615 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67989 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69418 "Found 1 new API test failure: /TestWebKit:WebKit.EvaluateJavaScriptThatCreatesBlob (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13463 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12221 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6405 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9814 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8146 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -445,26 +446,38 @@ static WTF::String getContentRuleListSourceFromMappedFile(const MappedData& mapp
     auto headerSizeBytes = headerSize(mappedData.metaData.version);
-    bool is8Bit = mappedData.data.span()[headerSizeBytes];
+
+    if (dataSpan.size() < headerSizeBytes + sourceSizeBytes) {
```
Should we also fail if it is greater than?

xeenon replied:
Greater than is valid still, since there is more data normally after the original source - if just the version number mismatches.

The other check for total file size catches the other case.

xeenon committed:
Crash at API::getContentRuleListSourceFromMappedFile().
https://webkit.org/b/283996
rdar://140450689

Reviewed by Alex Christensen.

Make the handling of getContentRuleListSourceFromMappedFile() more robust by checking header sizes
to the mapped data size. This was failing when the version got bumped and fallback to recompiling
from the original source was needed.

Improved mismatch version testing by dropping the version by one and keeping all the compiled data
the same to test proper version upgrades from the original source.

Added two corrupt size tests to error on invalid data sizes or empty data (missing source).

* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::getContentRuleListSourceFromMappedFile): Added logging and size checks.
(API::ContentRuleListStore::lookupContentRuleListFile): Check fileSize() against the mapped data.
(API::ContentRuleListStore::invalidateContentRuleListVersion): Changed to set the version back by one.
(API::ContentRuleListStore::corruptContentRuleListHeader): Added.
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStore.mm:
(-[WKContentRuleListStore _corruptContentRuleListHeaderForIdentifier:usingCurrentVersion:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStorePrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentExtensionStore.mm:
(TEST_F(WKContentRuleListStoreTest, VersionMismatch)): Updated expectations since source is preserved now.
(TEST_F(WKContentRuleListStoreTest, CorruptHeaderEmpty)): Added.
(TEST_F(WKContentRuleListStoreTest, CorruptHeaderRandom)): Added.

Canonical link: https://commits.webkit.org/287319@main
